src/models/geograph.py

```diff
@@ -614,21 +669,317 @@ def get_graph_components(
+            "valid_classes": self.valid_classes,
+            "barrier_classes": self.barrier_classes,
+            "max_travel_distance": self.max_travel_distance,
+            "add_distance": self.add_distance,
+        }
+        if save_path.suffix == ".bz2":
+            with bz2.BZ2File(save_path, "wb") as bz2_file:
+                pickle.dump(data, bz2_file)
+        elif save_path.suffix == ".gz":
+            with gzip.GzipFile(save_path, "wb") as gz_file:
+                gz_file.write(pickle.dumps(data))
+        else:
+            with open(save_path, "wb") as file:
+                pickle.dump(data, file)
+        save_path.chmod(0o664)
+
+
+class ComponentGeoGraph(GeoGraph):
+    """Class to represent the connected components of a GeoGraph."""
+
+    def __init__(
+        self,
+        components_list: List[set],
+        df: Optional[gpd.GeoDataFrame] = None,
+        add_distance_edges: bool = False,
+    ) -> None:
+        """
+        Class for the connected components of a GeoGraph.
+
+        This class can load a graph from only a list of components, in which
+        case it will not have a dataframe and the `has_df` class attribute will
+        be False, or it can load from both a list of components and a dataframe.
+        In the latter case, edges with the polygon distance can be added between
+        each pair of nodes (where each node corresponds to a connected component
+        in the original graph).
+
+        WARNING: using `add_distance_edges=True` is incredibly slow for anything
+        other than the smallest habitats.
+
+        Args:
+            components_list (List[set]): A list of sets, where each set contains
+            the node indices making up a component from the original graph.
+            df (Optional[gpd.GeoDataFrame], optional): A GeoDataFrame in which
+            each row contains a polygon that represents a component from a
+            GeoGraph. This is optional, and if not passed the graph will be
+            created with no edges. Defaults to None.
+            add_distance_edges (bool, optional): Boolean that determines whether
+            to add edges between every pair of nodes, with the distance between
+            their corresponding polygons as an edge attribute. Defaults to False.
+        """
+        # pylint: disable=super-init-not-called
+        self.has_df: bool = True
+        self.graph: nx.Graph = nx.Graph()
+        self.components_list: List[set] = components_list
+        if df is not None:
+            self.has_distance_edges: bool = add_distance_edges
+            self.df = self._load_from_dataframe(df, add_distance_edges)
+        else:
+            self.has_df = False
+            self.has_distance_edges = False
+            self.graph = nx.empty_graph(len(self.components_list))
+        self.metrics: Dict[str, Metric] = {}
+
+    def _load_from_dataframe(
+        self,
+        df: gpd.GeoDataFrame,
+        tolerance: float = 0.0,
```

Comment:
What's the best-practice in this case actually? Is it to keep redundant arguments running (e.g. the way you did or with kwargs) or is it to remove them? I don't know but would be curious to learn

Reply:
I think it's to create the redundant arguments so that if people accidentally copy a function call from the superclass and apply it to the child class, it all still works. See https://github.com/python/mypy/issues/1237 https://en.wikipedia.org/wiki/Liskov_substitution_principle